utils/zerovec/src/ule/unvalidated.rs

```diff
@@ -0,0 +1,230 @@
+/// human-readable formats like JSON.
+///
+/// # Examples
+///
+/// Using an [`UnvalidatedStr`] as the key of a [`ZeroMap`]:
+///
+/// ```
+/// use zerovec::ZeroMap;
+/// use zerovec::ule::UnvalidatedStr;
+///
+/// let map: ZeroMap<UnvalidatedStr, usize> = [
+///     (UnvalidatedStr::from_str("abc"), 11),
+///     (UnvalidatedStr::from_str("def"), 22),
+///     (UnvalidatedStr::from_str("ghi"), 33)
+/// ].into_iter().collect();
+///
+/// let key = "abc";
+/// let value = map.get_copied_by(|uvstr| uvstr.0.cmp(key.as_bytes()));
+/// assert_eq!(Some(11), value);
+/// ```
+///
+/// [`ZeroMap`]: crate::ZeroMap
+#[repr(transparent)]
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+#[allow(clippy::exhaustive_structs)] // transparent newtype
+pub struct UnvalidatedStr(pub [u8]);
+
+impl fmt::Debug for UnvalidatedStr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // Debug as a string if possible
+        match self.try_as_str() {
+            Ok(s) => fmt::Debug::fmt(s, f),
+            Err(_) => fmt::Debug::fmt(&self.0, f),
+        }
+    }
+}
+
+impl UnvalidatedStr {
+    /// Create a [`UnvalidatedStr`] from a byte slice.
+    #[inline]
+    pub const fn from_byte_slice(other: &[u8]) -> &Self {
+        // Safety: UnvalidatedStr is transparent over [u8]
+        unsafe { core::mem::transmute(other) }
+    }
+
+    /// Create a [`UnvalidatedStr`] from a string slice.
+    #[inline]
+    pub const fn from_str(s: &str) -> &Self {
+        Self::from_byte_slice(s.as_bytes())
+    }
+
+    /// Create a [`UnvalidatedStr`] from boxed bytes.
+    #[inline]
+    pub fn from_boxed_bytes(other: Box<[u8]>) -> Box<Self> {
+        // Safety: UnvalidatedStr is transparent over [u8]
+        unsafe { core::mem::transmute(other) }
+    }
+
+    /// Get the bytes from a [`UnvalidatedStr].
+    #[inline]
+    pub const fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+
+    /// Attempt to convert a [`UnvalidatedStr`] to a `str`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use zerovec::ule::UnvalidatedStr;
+    ///
+    /// static a: &UnvalidatedStr = UnvalidatedStr::from_byte_slice(b"abc");
+    ///
+    /// let b = a.try_as_str().unwrap();
+    /// assert_eq!(b, "abc");
+    /// ```
+    // Note: this is const starting in 1.63
+    #[inline]
+    pub fn try_as_str(&self) -> Result<&str, core::str::Utf8Error> {
+        core::str::from_utf8(&self.0)
+    }
+}
+
+impl<'a> From<&'a [u8]> for &'a UnvalidatedStr {
+    #[inline]
+    fn from(other: &'a [u8]) -> Self {
+        UnvalidatedStr::from_byte_slice(other)
+    }
+}
+
+impl From<Box<[u8]>> for Box<UnvalidatedStr> {
+    #[inline]
+    fn from(other: Box<[u8]>) -> Self {
+        UnvalidatedStr::from_boxed_bytes(other)
+    }
+}
+
+impl<'a, const N: usize> From<&'a [u8; N]> for &'a UnvalidatedStr {
+    #[inline]
+    fn from(other: &'a [u8; N]) -> Self {
+        UnvalidatedStr::from_byte_slice(other)
+    }
+}
```

Comment:
I don't think `From<{&[u8], Box<[u8]>, &[u8; N]}>` should exist. As pointed out in my comments above, outside deserialization we should not have to do `[u8] -> UnvalidatedStr`, and this makes that too easy. I'd also remove `from_boxed_bytes`, as that's only used by deserialization, whereas `from_bytes_slice` is at least somewhat useful for documentation and testing.

Reply:
I removed the `From` impls involving bytes, since your point about encouraging the sources to be strings is valid. I kept the explicit function versions though because removing them removes functionality and you need to be intentional to use them.